wpcom-sitemap.php

```diff
@@ -0,0 +1,504 @@
+			} else {
+				$url['changefreq'] = 'monthly';
+			}
+			wpcom_sitemap_array_to_simplexml( array( 'url' => apply_filters( 'sitemap_url', $url, $post->ID ) ), $tree );
+			unset( $url );
+		}
+		$blog_home = array(
+			'loc' => esc_url( get_option( 'home' ) ),
+			'changefreq' => 'daily',
+			'priority' => '1.0'
+		);
+		if ( !empty( $latest_mod ) ) {
+			$blog_home['lastmod'] = w3cdate_from_mysql($latest_mod);
+			header( 'Last-Modified:' . mysql2date('D, d M Y H:i:s', $latest_mod, 0).' GMT' );
+		}
+		wpcom_sitemap_array_to_simplexml( array( 'url'=> apply_filters( 'sitemap_url_home', $blog_home ) ), $tree );
+		unset( $blog_home );
+
+		$tree = apply_filters( 'wpcom_print_sitemap', $tree, $latest_mod );
+
+		$xml = $tree->asXML();
+		unset($tree);
+		if ( !empty($xml) ) {
+			wp_cache_set( $key, $xml, 'sitemap', 24*60*60 );  // cache for 24 hours
+			echo $xml;
+		}
+	} else {
+		header('Content-Type: ' . sitemap_content_type(), True);
+		echo $xml;
+	}
+	die();
+}
+
+function wpcom_print_news_sitemap($format) {
+	if ( defined( 'WPCOM_SKIP_DEFAULT_NEWS_SITEMAP' ) && WPCOM_SKIP_DEFAULT_NEWS_SITEMAP )
+		return;
+
+	global $wpdb;
+	$post_types = apply_filters( 'wpcom_sitemap_news_sitemap_post_types', array( 'post' ) );
+	if ( empty( $post_types ) )
+		return;
+
+	$post_types_in = array();
+	foreach ( $post_types as $post_type ) {
+		$post_types_in[] = $wpdb->prepare( '%s', $post_type );
+	}
+	$post_types_in_string = implode( ', ', $post_types_in );
+
+	$limit = apply_filters( 'wpcom_sitemap_news_sitemap_count', 1000 );
+	$cur_datetime = current_time( 'mysql', true );
+
+	$query = $wpdb->prepare( "
+		SELECT p.ID, p.post_title, p.post_type, p.post_date, p.post_name, p.post_date_gmt, GROUP_CONCAT(t.name SEPARATOR ', ') AS keywords
+		FROM
+			$wpdb->posts AS p LEFT JOIN $wpdb->term_relationships AS r ON p.ID = r.object_id
+			LEFT JOIN $wpdb->term_taxonomy AS tt ON r.term_taxonomy_id = tt.term_taxonomy_id AND tt.taxonomy = 'post_tag'
+			LEFT JOIN $wpdb->terms AS t ON tt.term_id = t.term_id
+		WHERE
+			post_status='publish' AND post_type IN ( {$post_types_in_string} ) AND post_date_gmt > (%s - INTERVAL 2 DAY)
+		GROUP BY p.ID
+		ORDER BY p.post_date_gmt DESC LIMIT %d", $cur_datetime, $limit );
+
+
+	header('Content-Type: application/xml');
+	echo '<?xml version="1.0" encoding="UTF-8"?>'."\n";
+?>
+<!-- generator="wordpress.com" -->
+<urlset xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd"
+	xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
+	xmlns:news="http://www.google.com/schemas/sitemap-news/0.9"
+	xmlns:image="http://www.google.com/schemas/sitemap-image/1.1"
+	>
+<?php
+	$posts = wp_cache_get('news_sitemap_posts', 'metro_sitemap');
+	if( empty( $posts ) ) {
+		$posts = $wpdb->get_results( $query );
+		wp_cache_set('news_sitemap_posts', $posts, 'metro_sitemap', 60);
```

Comment:
Same as above.  Can we also add a filter for the cache time?

Reply:
I've merged this so we can get it out to all sites, but let's discuss the filter in a new issue.